src/ClientData/ThreadTrackPaneData.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <ClientData/ThreadTrackPaneData.h>
+
+namespace orbit_client_data {
+
+size_t ThreadTrackPaneData::GetNumberOfTimers() const {
+  absl::MutexLock lock(&scope_tree_mutex_);
+  // Special case. ScopeTree has a root node in depth 0 which shouldn't be considered.
+  return scope_tree_.Size() - 1;
+}
+
+uint32_t ThreadTrackPaneData::GetMaxDepth() const {
+  absl::MutexLock lock(&scope_tree_mutex_);
+  return scope_tree_.Height() - 1;
+}
+
+const orbit_client_protos::TimerInfo& ThreadTrackPaneData::AddTimer(
+    orbit_client_protos::TimerInfo timer_info) {
+  const auto& timer_info_chain_ref = track_pane_data_.AddTimer(/*depth=*/0, timer_info);
+  UpdateMinTime(timer_info.start());
+  UpdateMaxTime(timer_info.end());
+
+  if (scope_tree_update_type_ == ScopeTreeUpdateType::kAlways) {
+    absl::MutexLock lock(&scope_tree_mutex_);
+    scope_tree_.Insert(&timer_info_chain_ref);
+  }
+  return timer_info_chain_ref;
+}
+
+void ThreadTrackPaneData::BuildScopeTreeFromTrackData() {
+  if (scope_tree_update_type_ != ScopeTreeUpdateType::kOnCaptureComplete) {
+    return;
+  }
+  std::vector<const TimerChain*> timer_chains = track_pane_data_.GetChains();
+  for (const TimerChain* timer_chain : timer_chains) {
+    CHECK(timer_chain != nullptr);
+    absl::MutexLock lock(&scope_tree_mutex_);
+    for (const auto& block : *timer_chain) {
+      for (size_t k = 0; k < block.size(); ++k) {
+        scope_tree_.Insert(&block[k]);
+      }
+    }
+  }
+}
+
+std::vector<const orbit_client_protos::TimerInfo*> ThreadTrackPaneData::GetTimers(
+    uint64_t min_tick, uint64_t max_tick) const {
+  std::vector<const orbit_client_protos::TimerInfo*> all_timers;
+  absl::MutexLock lock(&scope_tree_mutex_);
+
+  for (const auto& [depth, ordered_nodes] : scope_tree_.GetOrderedNodesByDepth()) {
+    // Special case. ScopeTree has a root node in depth 0 which shouldn't be considered.
+    if (!depth) continue;
+
+    auto first_node_to_draw = ordered_nodes.upper_bound(min_tick);
+    if (first_node_to_draw != ordered_nodes.begin()) --first_node_to_draw;
+
+    // If this node is strictly before the range, we shouldn't include it.
+    if (first_node_to_draw->second->GetScope()->end() < min_tick) ++first_node_to_draw;
```

Comment:
Does this happen? Wouldn't it mean that the end time is the same as the start time and the same as min_time?

Reply:
I've added this line because my tests didn't work with the original code. I think that it's a current bug in ThreadTrack.cpp but anyway the additional timer is being filter later, so not a problem at all in that file. I will try to explain it with an example.

Image, we are checking (4,200) and in the ScopeTree we have (1,3) (10,20) (300,400). 
The original code (without this line) points as (1,3) to the first_node_to_draw. Why? Because as we are making a lower_bound/upper_bound by start time and after erasing 1, we will point to the previous node, so the children in the ScopeTree who has starting_point less (or equal if we use upper_bound) than the query_start. 

But actually, this is only going to be the first_node_to_draw only if the ending_time is inside the query_range. So, we need to add this if. 

Again, the code in ThreadTrack.cpp is working anyway because this timer (1,3) is being filtered later.

I thought a little and I didn't have a better solution, but I'm open to suggestions :)